src/analysis/convex/topology.lean

```diff
@@ -0,0 +1,183 @@
+
+local attribute [instance] set.pointwise_add set.smul_set
+
+/-- In a topological vector space, the interior of a convex set is convex. -/
+lemma convex.interior {s : set E} (hs : convex s) : convex (interior s) :=
+convex_iff_set.mpr $ λ a b ha hb hab,
+  have h : is_open (a • interior s + b • interior s), from
+  or.elim (classical.em (a = 0))
+  (λ heq,
+    have hne : b ≠ 0, by { rw [heq, zero_add] at hab, rw hab, exact one_ne_zero },
+    is_open_pointwise_add_left ((is_open_map_smul_of_ne_zero hne _) is_open_interior))
+  (λ hne,
+    is_open_pointwise_add_right ((is_open_map_smul_of_ne_zero hne _) is_open_interior)),
+  (subset_interior_iff_subset_of_open h).mpr $ subset.trans
+    (by { apply pointwise_add_subset_add; exact image_subset _ interior_subset })
+    (convex_iff_set.mp hs ha hb hab)
+
+/-- In a topological vector space, the closure of a convex set is convex. -/
+lemma convex.closure {s : set E} (hs : convex s) : convex (closure s) :=
+λ x y hx hy a b ha hb hab,
+let f : E → E → E := λ x' y', a • x' + b • y' in
+have hf : continuous (λ p : E × E, f p.1 p.2), from
+  (continuous_const.smul continuous_fst).add (continuous_const.smul continuous_snd),
+show f x y ∈ closure s, from
+  mem_closure_of_continuous2 hf hx hy (λ x' hx' y' hy', subset_closure
+  (hs hx' hy' ha hb hab))
+
+/-- Convex hull of a finite set is compact. -/
+lemma set.finite.compact_convex_hull {s : set E} (hs : finite s) :
+  compact (convex_hull s) :=
+begin
+  rw [hs.convex_hull_eq_image],
+  apply (compact_std_simplex _).image,
+  haveI := hs.fintype,
+  apply linear_map.continuous_on_pi
+end
+
+/-- Convex hull of a finite set is closed. -/
+lemma set.finite.is_closed_convex_hull [t2_space E] {s : set E} (hs : finite s) :
+  is_closed (convex_hull s) :=
+closed_of_compact _ hs.compact_convex_hull
+
+end topological_vector_space
+
+/-! ### Normed vector space -/
+
+section normed_space
+variables [normed_group E] [normed_space ℝ E]
+
+lemma convex_on_dist (z : E) (s : set E) (hs : convex s) :
+  convex_on s (λz', dist z' z) :=
+and.intro hs $
+assume x y hx hy a b ha hb hab,
+calc
+  dist (a • x + b • y) z = ∥ (a • x + b • y) - (a + b) • z ∥ :
+    by rw [hab, one_smul, normed_group.dist_eq]
+  ... = ∥a • (x - z) + b • (y - z)∥ :
+    by rw [add_smul, smul_sub, smul_sub]; simp
+  ... ≤ ∥a • (x - z)∥ + ∥b • (y - z)∥ :
+    norm_add_le (a • (x - z)) (b • (y - z))
+  ... = a * dist x z + b * dist y z :
+    by simp [norm_smul, normed_group.dist_eq, real.norm_eq_abs, abs_of_nonneg ha, abs_of_nonneg hb]
+
+lemma convex_ball (a : E) (r : ℝ) : convex (metric.ball a r) :=
+by simpa only [metric.ball, sep_univ] using (convex_on_dist a _ convex_univ).convex_lt r
+
+lemma convex_closed_ball (a : E) (r : ℝ) : convex (metric.closed_ball a r) :=
+by simpa only [metric.closed_ball, sep_univ] using (convex_on_dist a _ convex_univ).convex_le r
+
+lemma convex_hull_exists_dist_ge {s : set E} {x : E} (hx : x ∈ convex_hull s) (y : E) :
```

Comment:
A docstring explaining the content of the lemma would be useful.

Reply:
Done.